File: src/repository/systemState.ts
import { db } from "../database/db.ts"

const SYSTEM_STATE_KEY_JETSTREAM_CURSOR = "jetstream_cursor"

export const getJetstreamCursor = async (): Promise<number | undefined> => {
  const result = await db.selectFrom("system_states")
    .select("value")
    .where("key", "=", SYSTEM_STATE_KEY_JETSTREAM_CURSOR)
    .executeTakeFirst()

  return result?.value
}

export const saveJetstreamCursor = async (cursor: number): Promise<void> => {
  await db.insertInto("system_states")
    .values({
      key: SYSTEM_STATE_KEY_JETSTREAM_CURSOR,
      value: cursor,
    })
    .onDuplicateKeyUpdate({
      value: cursor,
    })
    .execute()
}

File: src/repository/system_state.rs
use crate::database::DbPool;
use crate::model::jetstream::JETSTREAM_CURSOR_KEY;

pub async fn get_jetstream_cursor(pool: &DbPool) -> anyhow::Result<Option<i64>> {
    let row = sqlx::query_scalar!(
        "SELECT value FROM system_states WHERE `key` = ?",
        JETSTREAM_CURSOR_KEY
    )
    .fetch_optional(pool)
    .await?;

    Ok(row)
}

pub async fn save_jetstream_cursor(pool: &DbPool, cursor: i64) -> anyhow::Result<()> {
    sqlx::query!(
        "INSERT INTO system_states (`key`, value) VALUES (?, ?) ON DUPLICATE KEY UPDATE value = ?",
        JETSTREAM_CURSOR_KEY,
        cursor,
        cursor
    )
    .execute(pool)
    .await?;

    Ok(())
}
